src/crl/interactivesessions/shells/remoteshell.py

```diff
@@ -0,0 +1,135 @@
+
+import logging
+import paramiko
+
+from .paramikospawn import ParamikoSpawn
+
+from crl.interactivesessions.interactivesessionexceptions import (
+    InteractiveSessionError)
+from crl.interactivesessions.shells.bashshell import BashShell
+from crl.interactivesessions.shells.sshoptions import sshoptions
+from crl.interactivesessions.shells.registershell import RegisterShell
+from crl.interactivesessions.shells.msgreader import MsgReader
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class SshError(InteractiveSessionError):
+    """
+    Raised when :class:`.RemoteShell` start fails
+    """
+
+
+@RegisterShell()
+class RemoteShell(BashShell):
+    """
+    This class can be used in order to start a remote bash shell. See also
+    :class:`.BashShell`.
+    **Args:**
+    *ip*: IP address of the host
+    *username*: login username
+    *password*: login passowrd. If not given, passwordless login is expected.
+    *tty_echo*: If True, then terminal echo is set on.
+    *second_password*: if not *None*, this password is send to terminal after
+                       the first password succesfully aplied.
+    *port*: If *port* is not None, alternate port is used in the connection
+            instead of the detfault 22.
+    *init_env*: Path to initialization file which is sourced after the all
+                other initialization is done.
+    *source*: File to source to initial the terminal at the ssh host
+    *sudo* If super user should be used for commands set this to true
+           Default is false. It will initialize shell session with sudo -i
+    For setting timeout for reading login banner, i.e. message-of-day, please use
+    :meth:`.msgreader.MsgReader.set_timeout`.
+    """
+    # TODO: add "-oLogLevel=error" to avoid banner...
+    _ssh_options = sshoptions
+
+    def __init__(self, ip, username=None, password=None, tty_echo=False,
+                 second_password=None, port=None, init_env=None, key_file=None,
+                 source=None, sudo=None):
+        super(RemoteShell, self).__init__(tty_echo=tty_echo)
+        self.ip = ip
+        self.username = username
+        self.passwords = [] if password is None else [password]
+        self.port = port
+        self.key_file = key_file
+        self.source = source
+        self.sudo = sudo
+        self.ssh = None
+        self.chan = None
+        self.start = self._paramikospawn
+        self.init_env = init_env
+        if second_password:
+            self.passwords.append(second_password)
+
+    def __getattr__(self, name):
+        if name == 'spawn':
+            return self._paramikospawn
+
+    def get_start_cmd(self):
+        ssh_options = RemoteShell._ssh_options
+        if self.port is not None:
+            ssh_options += ' -p {}'.format(int(self.port))
+
+        return ("ssh {0} {1}".format(ssh_options, self.ip)
+                if self.username is None else
+                "ssh {0} {1}@{2}".format(ssh_options, self.username, self.ip))
```

Comment:
these are SSH client tool specific. This is paramiko based. We likely do not need these.